File: ios/NativeSigner/Cards/HistoryCardExtended.swift
//
//  HistoryCardExtended.swift
//  NativeSigner
//
//  Created by Alexander Slesarev on 28.10.2021.
//

import SwiftUI

struct HistoryCardExtended: View {
    var event: MEventMaybeDecoded
    var body: some View {
        switch event.event {
        case .databaseInitiated,
             .deviceWasOnline,
             .generalVerifierSet,
             .historyCleared,
             .identitiesWiped,
             .metadataAdded,
             .metadataRemoved,
             .networkSpecsAdded,
             .networkSpecsRemoved,
             .resetDangerRecord,
             .seedCreated,
             .seedRemoved,
             .seedNameWasShown,
             .networkSpecsSigned,
             .systemEntry,
             .typesAdded,
             .typesRemoved,
             .userEntry,
             .warning,
             .wrongPassword,
             .messageSignError,
             .messageSigned:
            HistoryCard(event: event.event)
        case let .identityAdded(value):
            HistoryCardTemplate(
                image: .init(.aqi, variant: .medium),
                danger: false,
                line1: Localizable.keyCreated.string,
                line2: value.seedName + value.path + " in network with hash " +
                    value.networkGenesisHash.formattedAsString
            )
        case let .identityRemoved(value):
            HistoryCardTemplate(
                image: .init(.xmark, variants: [.rectangle, .portrait]),
                danger: false,
                line1: Localizable.keyRemoved.string,
                line2: value.seedName + value.path + " in network with hash " +
                    value.networkGenesisHash.formattedAsString
            )
        case let .secretWasExported(value):
            HistoryCardTemplate(
                image: .init(.eye, variants: [.trianglebadge, .exclamationmark, .fill]),
                danger: true,
                line1: Localizable.secretWasExported.string,
                line2: value.seedName + value.path + " in network with hash " +
                    value.networkGenesisHash.formattedAsString
            )
        case let .networkVerifierSet(value):
            switch value.validCurrentVerifier {
            case .general:
                HistoryCardTemplate(
                    image: .init(.checkmark, variant: .shield),
                    danger: false,
                    line1: Localizable.networkVerifierSet.string,
                    line2: value.generalVerifier.show() + " for network with genesis hash " +
                        value.genesisHash.formattedAsString
                )
            case let .custom(verifier):
                HistoryCardTemplate(
                    image: .init(.checkmark, variant: .shield),
                    danger: false,
                    line1: Localizable.networkVerifierSet.string,
                    line2: verifier.show() + " for network with genesis hash " +
                        value.genesisHash.formattedAsString
                )
            }
        case let .metadataSigned(value):
            HistoryCardTemplate(
                image: .init(.signature),
                danger: false,
                line1: Localizable.metadataSigned.string,
                line2: value.name + String(value.version)
            )
        case let .typesSigned(value):
            HistoryCardTemplate(
                image: .init(.signature),
                danger: false,
                line1: Localizable.typesSigned.string,
                line2: value.typesHash.formattedAsString
            )
        case let .transactionSignError(value):
            VStack {
                Localizable.transactionFailed.text
                OldTransactionBlock(cards: event.decoded?.asSortedCards() ?? [])
                Localizable.signedBy.text
                HStack {
                    Identicon(identicon: event.signedBy?.address.identicon.svgPayload ?? [])
                    VStack {
                        Text(value.signedBy.show())
                        Text((event.signedBy?.address.seedName ?? "") + (event.signedBy?.address.path ?? ""))
                    }
                }
                Localizable.inNetwork.text
                Text(value.networkName)
                Localizable.commentAlt.text
                Text(value.userComment)
            }
        case let .transactionSigned(value):
            VStack {
                OldTransactionBlock(cards: event.decoded?.asSortedCards() ?? [])
                Localizable.signedBy.text
                HStack {
                    Identicon(identicon: event.signedBy?.address.identicon.svgPayload ?? [])
                    VStack {
                        Text(value.signedBy.show())
                        Text((event.signedBy?.address.seedName ?? "") + (event.signedBy?.address.path ?? ""))
                    }
                }
                Localizable.inNetwork.text
                Text(value.networkName)
                Localizable.commentAlt.text
                Text(value.userComment)
            }
        }
    }
}

struct OldTransactionBlock: View {
    var cards: [TransactionCard]
    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: CornerRadius.extraSmall)
                .stroke(Asset.crypto400.swiftUIColor)
            VStack {
                ForEach(cards, id: \.index) { card in
                    TransactionCardView(card: card)
                }
            }
            .padding(Spacing.medium)
        }
        .padding(Spacing.small)
        .frame(width: UIScreen.main.bounds.size.width)
    }
}

// struct HistoryCardExtended_Previews: PreviewProvider {
// static var previews: some View {
// HistoryCardExtended()
// }
// }

File: android/app/src/main/java/io/parity/signer/components/HistoryCardExtended.kt
package io.parity.signer.components

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.runtime.Composable
import io.parity.signer.models.abbreviateString
import io.parity.signer.uniffi.Event
import io.parity.signer.uniffi.MEventMaybeDecoded
import io.parity.signer.uniffi.VerifierValue

/**
 * Detailed history event description representation selector
 */
@Composable
fun HistoryCardExtended(
	event: MEventMaybeDecoded,
	timestamp: String
) {
	val decodedTransaction = event.decoded
	val signedBy = event.signedBy
	val verifierDetails = event.verifierDetails
	when (val eventVal = event.event) {
		is Event.DatabaseInitiated -> {
			HistoryCardTemplate(
				image = Icons.Default.Smartphone,
				line1 = timestamp,
				line2 = "Database initiated",
				line3 = ""
			)
		}
		is Event.DeviceWasOnline -> {
			HistoryCardTemplate(
				image = Icons.Default.Dangerous,
				line1 = timestamp,
				line2 = "Device was connected to network",
				line3 = "",
				danger = true
			)
		}
		is Event.GeneralVerifierSet -> {
			val hex = eventVal.verifier.v.let {
				when (it) {
					is VerifierValue.Standard -> {
						it.m
					}
					else -> listOf()
				}
			}
			HistoryCardTemplate(
				image = Icons.Default.Shield,
				line1 = timestamp,
				line2 = "General verifier set",
				line3 = hex.getOrElse(0) { "" }
					.abbreviateString(8) + hex.getOrElse(1) { "" }
			)
		}
		is Event.HistoryCleared -> {
			HistoryCardTemplate(
				image = Icons.Default.DeleteForever,
				line1 = timestamp,
				line2 = "History cleared",
				line3 = ""
			)
		}
		is Event.IdentitiesWiped -> {
			HistoryCardTemplate(
				image = Icons.Default.Delete,
				line1 = timestamp,
				line2 = "All keys were wiped",
				line3 = ""
			)
		}
		is Event.IdentityAdded -> {
			eventVal.identityHistory.let {
				HistoryCardTemplate(
					image = Icons.Default.Pattern,
					line1 = timestamp,
					line2 = "Key created",
					line3 = it.seedName + it.path
				)
			}
		}
		is Event.IdentityRemoved -> {
			eventVal.identityHistory.let {
				HistoryCardTemplate(
					image = Icons.Default.Delete,
					line1 = timestamp,
					line2 = "Key removed",
					line3 = it.seedName + it.path
				)
			}
		}
		is Event.MessageSignError -> {
			HistoryCardTemplate(
				image = Icons.Default.Warning,
				line1 = timestamp,
				line2 = "Message signing error!",
				line3 = "message:" + eventVal.signMessageDisplay.message + " user comment: " + eventVal.signMessageDisplay.userComment,
				danger = true
			)
		}
		is Event.MessageSigned -> {
			HistoryCardTemplate(
				image = Icons.Default.Done,
				line1 = timestamp,
				line2 = "Generated signature for message",
				line3 = "message:" + eventVal.signMessageDisplay.message + " user comment: " + eventVal.signMessageDisplay.userComment
			)
		}
		is Event.MetadataAdded -> {
			eventVal.metaValuesDisplay.let {
				HistoryCardTemplate(
					image = Icons.Default.QrCodeScanner,
					line1 = timestamp,
					line2 = "Metadata added",
					line3 = it.name + " version " + it.version
				)
			}
		}
		is Event.MetadataRemoved -> {
			eventVal.metaValuesDisplay.let {
				HistoryCardTemplate(
					image = Icons.Default.Delete,
					line1 = timestamp,
					line2 = "Metadata removed",
					line3 = it.name + " version " + it.version
				)
			}
		}
		is Event.NetworkSpecsAdded -> {
			HistoryCardTemplate(
				image = Icons.Default.QrCodeScanner,
				line1 = timestamp,
				line2 = "Network added",
				line3 = eventVal.networkSpecsDisplay.specs.title,
			)
		}
		is Event.NetworkSpecsRemoved -> {
			HistoryCardTemplate(
				image = Icons.Default.Delete,
				line1 = timestamp,
				line2 = "Network removed",
				line3 = eventVal.networkSpecsDisplay.specs.title
			)
		}
		is Event.NetworkVerifierSet -> {
			HistoryCardTemplate(
				image = Icons.Default.Shield,
				line1 = timestamp,
				line2 = "Network verifier set",
				line3 = eventVal.networkVerifierDisplay.genesisHash.toString()
			)
		}
		is Event.ResetDangerRecord -> {
			HistoryCardTemplate(
				image = Icons.Default.DeleteForever,
				line1 = timestamp,
				line2 = "History cleared",
				line3 = "",
				danger = true
			)
		}
		is Event.SeedCreated -> {
			HistoryCardTemplate(
				image = Icons.Default.Pattern,
				line1 = timestamp,
				line2 = "Seed created",
				line3 = eventVal.seedCreated
			)
		}
		is Event.SeedNameWasShown -> {
			HistoryCardTemplate(
				image = Icons.Default.Warning,
				line1 = timestamp,
				line2 = "Seed was shown",
				line3 = eventVal.seedNameWasShown
			)
		}
		is Event.NetworkSpecsSigned -> {
			HistoryCardTemplate(
				image = Icons.Default.Verified,
				line1 = timestamp,
				line2 = "Network specs signed",
				line3 = eventVal.networkSpecsExport.specsToSend.title
			)
		}
		is Event.MetadataSigned -> {
			eventVal.metaValuesExport.let {
				HistoryCardTemplate(
					image = Icons.Default.Verified,
					line1 = timestamp,
					line2 = "Meta signed",
					line3 = it.name + it.version
				)
			}
		}
		is Event.TypesSigned -> {
			HistoryCardTemplate(
				image = Icons.Default.Verified,
				line1 = timestamp,
				line2 = "Types signed",
				line3 = ""
			)
		}
		is Event.SystemEntry -> {
			HistoryCardTemplate(
				image = Icons.Default.Warning,
				line1 = timestamp,
				line2 = "System entry",
				line3 = eventVal.systemEntry
			)
		}
		is Event.TransactionSignError -> {
			HistoryCardTemplate(
				image = Icons.Default.Dangerous,
				line1 = timestamp,
				line2 = "Signing failure",
				line3 = eventVal.signDisplay.userComment,
				danger = true
			)
		}
		is Event.TransactionSigned -> {
			Column {
				Text("Transaction signed")

				if (decodedTransaction != null) {
					TransactionPreviewField(
						cardSet = decodedTransaction
					)
				}
				Text("Signed by:")
				Row {
					Identicon(
						identicon = signedBy?.identicon ?: listOf()
					)
					Column {
						Text(verifierDetails?.publicKey ?: "")
						Text(
							verifierDetails?.encryption ?: ""
						)
					}
				}
				Text("In network")
				Text(eventVal.signDisplay.networkName)
				Text("Comment:")
				Text(
					eventVal.signDisplay.userComment
				)
			}
		}
		is Event.TypesAdded -> {
			HistoryCardTemplate(
				image = Icons.Default.QrCodeScanner,
				line1 = timestamp,
				line2 = "New types info loaded",
				line3 = ""
			)
		}
		is Event.TypesRemoved -> {
			HistoryCardTemplate(
				image = Icons.Default.Remove,
				line1 = timestamp,
				line2 = "Types info removed",
				line3 = "",
				danger = true
			)
		}
		is Event.UserEntry -> {
			HistoryCardTemplate(
				image = Icons.Default.Note,
				line1 = timestamp,
				line2 = "User entry",
				line3 = eventVal.userEntry
			)
		}
		is Event.Warning -> {
			HistoryCardTemplate(
				image = Icons.Default.Warning,
				line1 = timestamp,
				line2 = "Warning!",
				line3 = eventVal.warning,
				danger = true
			)
		}
		is Event.WrongPassword -> {
			HistoryCardTemplate(
				image = Icons.Default.Warning,
				line1 = timestamp,
				line2 = "Wrong password entered",
				line3 = "operation declined",
				danger = true
			)
		}
	}
}
